Add elasticsearch-node detach-cluster tool

andrershov opened the pull request:
This commit adds the second part of `elasticsearch-node` tool -
`detach-cluster` command in addition to `unsafe-bootstrap` command.
Also, this commit changes the semantics of `unsafe-bootstrap`, now
`unsafe-bootstrap` changes clusterUUID.
So the algorithm of running `elasticsearch-node` tool is the following:
1) Stop all nodes in the cluster.
2) Pick master-eligible node with the highest (term, version) pair and
run the `unsafe-bootstrap` command on it. If there are no survived
master-eligible nodes - skip this step.
3) Run `detach-cluster` command on the remaining survived nodes.

Detach cluster makes the following changes to the node metadata:
1) Sets clusterUUID commited to false.
2) Sets currentTerm and term to 0. 
3) Removes voting tombstones and sets voting configurations to special
constant MUST_JOIN_ELECTED_MASTER, that prevents initial cluster
bootstrap.

`ElasticsearchNodeCommand` base abstract class is introduced, because
`UnsafeBootstrapMasterCommand` and `DetachClusterCommand` have a lot in
common.
Also, this commit adds "ordinal" parameter to both commands, because it's 
impossible to write IT otherwise.
For MUST_JOIN_ELECTED_MASTER case special handling is introduced in
`ClusterFormationFailureHelper`.
Tests for both commands reside in `ElasticsearchNodeCommandIT` (renamed
from `UnsafeBootstrapMasterIT`).

elasticmachine commented:
Pinging @elastic/es-distributed

andrershov commented on server/src/main/java/org/elasticsearch/cluster/coordination/DetachClusterCommand.java:
```diff
@@ -0,0 +1,79 @@
+            "--------------------------------------------------------------------------\n" +
+                    "\n" +
+                    "You should run this tool only if you have permanently lost all\n" +
+                    "your master-eligible nodes, and you cannot restore the cluster\n" +
```
See below "or you have already run `elasticsearch-node unsafe-bootstrap` ...". Probably @DaveCTurner will have to come up with better wording.

andrershov commented on server/src/main/java/org/elasticsearch/cluster/coordination/DetachClusterCommand.java:
```diff
@@ -0,0 +1,79 @@
+                    "your master-eligible nodes, and you cannot restore the cluster\n" +
+                    "from a snapshot, or you have already run `elasticsearch-node unsafe-bootstrap`\n" +
+                    "on master-eligible node that formed cluster with this node.\n" +
+                    "This tool can result in arbitrary data loss and should be\n" +
```
This sentence is just copied from unsafe-bootstrap command

ywelsch replied:
sure, that still doesn't make it great :)

andrershov commented:
@ywelsch Thank you for the review!
I've addressed your comments, except wording.
Regarding tests, the two tests that you meant are `testDanglingIndices` and `testIndexImportedFromDataOnlyNodesIfMasterLostDataFolder` in my opinion they are the same and one of them should be removed.
Taking this into account, I've implemented `ElasticsearchNodeCommandIT.testAllMasterEligibleNodesFailedDanglingIndexImport` which replaces both  `testDanglingIndices` and `testIndexImportedFromDataOnlyNodesIfMasterLostDataFolder`.
I think, `ElasticsearchNodeCommandIT` is a better place for this test because it requires `detachCluster` invocation and `ElasticsearchNodeCommandIT` is specifically designed for this case and makes `detachCluster` command execution one-liner. 
Please let me know what do you think.

andrershov commented:
@DaveCTurner thanks for your review! I've addressed `ClusterFormationFailureHelper` message in https://github.com/elastic/elasticsearch/pull/37979/commits/a122221a0986982a692c0eb2dc673d93cc001365 and all wording in https://github.com/elastic/elasticsearch/pull/37979/commits/80ad2248d1469b1ae602e891e91f63c594076734.
@DaveCTurner and @ywelsch it's ready for the next pass.

andrershov commented:
run elasticsearch-ci/2

andrershov commented:
run elasticsearch-ci/2

andrershov commented:
run elasticsearch-ci/2

colings86 commented:
Please remember to add all relevant labels (area label, version label(s) and change type label) on all PRs and please look for this as part of reviews. The release note generation process is made much harder when PRs are not labelled correctly.
